### PR TITLE
config to disable randomness in a chain halt

### PR DESCRIPTION
**Pull request by zjma**
## Description

Adding:
- a local config to disable randomness at a chain halt caused by randomness generation failure,
- an on-chain config to re-enable randomness after the recovery.

## Type of Change
- [x] New feature

## Which Components or Systems Does This Change Impact?
- [x] Validator Node
- [x] Aptos Framework

## How Has This Been Tested?
new smoke test.

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/configs/randomness_config_seqnum.move [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+/// 1. Once the bug is fixed and the binary + framework have been patched,
+///    a governance proposal is needed to set `RandomnessConfigSeqNum` to be `X+2`.
+module aptos_framework::randomness_config_seqnum {
+    use aptos_framework::config_buffer;
+    use aptos_framework::system_addresses;
+
+    friend aptos_framework::reconfiguration_with_dkg;
+
+    /// If this seqnum is smaller than a validator local override, the on-chain `RandomnessConfig` will be ignored.
+    /// Useful when performing an emergency randomness
```
Comment (danielxiangzl) [inline aptos-move/framework/aptos-framework/sources/configs/randomness_config_seqnum.move:17]: emergency stop of randomness?

File: consensus/src/epoch_manager.rs [hunk 2]
```diff
@@ -1068,9 +1073,24 @@ impl<P: OnChainConfigProvider> EpochManager<P> {
...
+            .unwrap_or_else(|_| RandomnessConfigSeqNum::default_if_missing());
+
+        info!("local_seqnum={}", self.randomness_override_seq_num);
+        info!(
+            "onchain_seqnum={}",
+            onchain_randomness_config_seq_num.seq_num
+        );
+        let onchain_randomness_config =
+            if onchain_randomness_config_seq_num.seq_num < self.randomness_override_seq_num {
+                info!("Randomness emergency stop.");
```
Comment (danielxiangzl) [inline consensus/src/epoch_manager.rs:1086]: Use error! to alert if validator accidentally overrides its local node config?

File: testsuite/smoke-test/src/randomness/randomness_stall_recovery.rs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+use aptos_rest_client::Client;
+use aptos_types::{on_chain_config::OnChainRandomnessConfig, randomness::PerBlockRandomness};
+use futures::future::join_all;
+use std::{
+    ops::Add,
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+/// Verify the correctness of DKG transcript and block-level randomness seed.
```
Comment (danielxiangzl) [inline testsuite/smoke-test/src/randomness/randomness_stall_recovery.rs:18]: update comment

**Review by danielxiangzl — APPROVED** [review]
Amazing!

File: consensus/src/rand/rand_gen/rand_manager.rs [hunk 4]
```diff
@@ -170,6 +171,12 @@ impl<S: TShare, D: TAugmentedData> RandManager<S, D> {
             .iter()
             .flat_map(|b| b.ordered_blocks.iter().map(|b3| b3.round()))
             .collect();
+        fail_point!("rand_manager::process_ready_blocks", |_| {
```
Comment (zekun000) [inline consensus/src/rand/rand_gen/rand_manager.rs:174]: is this change related?

File: dkg/src/epoch_manager.rs [hunk 5]
```diff
@@ -162,10 +167,25 @@ impl<P: OnChainConfigProvider> EpochManager<P> {
...
+        let onchain_randomness_config =
+            if onchain_randomness_config_seq_num.seq_num < self.randomness_override_seq_num {
+                info!("Randomness emergency stop.");
+                OnChainRandomnessConfig::default_disabled()
+            } else {
+                payload
+                    .get::<RandomnessConfigMoveStruct>()
+                    .and_then(OnChainRandomnessConfig::try_from)
+                    .unwrap_or_else(|_| OnChainRandomnessConfig::default_if_missing())
+            };
```
Comment (zekun000) [inline dkg/src/epoch_manager.rs:188]: feels like this code block can be reused, adding a function in the OnChainRandomnessConfig to take payload and local seqnum as params?

File: aptos-move/framework/aptos-framework/sources/configs/randomness_config_seqnum.move (hunk 1 above)
Comment (zjma) [reply aptos-move/framework/aptos-framework/sources/configs/randomness_config_seqnum.move]: fixed

File: consensus/src/epoch_manager.rs (hunk 2 above)
Comment (zjma) [reply consensus/src/epoch_manager.rs]: warned instead

File: testsuite/smoke-test/src/randomness/randomness_stall_recovery.rs (hunk 3 above)
Comment (zjma) [reply testsuite/smoke-test/src/randomness/randomness_stall_recovery.rs]: fixed

File: consensus/src/rand/rand_gen/rand_manager.rs (hunk 4 above)
Comment (zjma) [reply consensus/src/rand/rand_gen/rand_manager.rs]: fixed

File: dkg/src/epoch_manager.rs (hunk 5 above)
Comment (zjma) [reply dkg/src/epoch_manager.rs]: fixed

File: consensus/src/rand/rand_gen/rand_manager.rs (hunk 4 above)
Comment (zjma) [reply consensus/src/rand/rand_gen/rand_manager.rs]: actually that's needed by the new smoke test. Bringing it back.

**Review by zekun000 — APPROVED** [review]
